Add advanced guide for congestion control to Sui docs (#20892)
MystenLabs/sui: docs/content/guides/developer/advanced/congestion-control.mdx

---

> title: Congestion Control

Reviewer: Can we call this "object-based local fee markets" (and use that term throughout)? This is the terminology I use when comparing/contrasting with how local fee markets work elsewhere, and it emphasizes the positive aspects of the system (static object info helps determine a per object budget up front!)

Author: Congestion control in my mind is an umbrella term that includes BOTH the fee markets (i.e. how do you prioritize tx relative to each other) AND the scheduling logic for determining transaction cost estimates & capacity, deferring tx that exceed budget, etc.

I am not sure it makes sense to me to use "fee markets" to refer to the whole system? If you think we still want to do that in order to make the comparison to other systems more obvious I can, but wanted to raise this point first - what do you think?

Reviewer: I think you're 100% right on "congestion control" being the technically accurate/broader term, and the motivation my request is mostly as you say

> If you think we still want to do that in order to make the comparison to other systems more obvious I can,

: external tech communication, and technical marketing. At a high level, I think we want to be able to say "Sui is great because it has X, Y and Z (e.g., object-centric data model, PTBs, Mysticeti, object-based local fee markets)", be very consistent about the terms for X, Y, Z, and expect folks to arrive at the docs looking to learn about these features.

I think X = "Sui has object-baed local fee markets" works well for this purpose (and have been using it), but "Sui has congestion control" is not as good (too easy to anchor on "Sui has congestion?"

description: Congestion control is the system that limits the rate of transactions writing to a single shared object, preventing the network from becoming overloaded with checkpoints that take too long to execute.
---

> The Sui network’s object-based architecture allows us to process many different user transactions massively in parallel, in a way that’s not possible on most other networks. However, if multiple transactions are all writing to the same shared object, they must execute in sequential order. There is a limit to how many transactions we can process that touch one specific object.

Reviewer 2: ```suggestion
The Sui network’s object-based architecture allows processing many different user transactions in parallel, in a way that’s not possible on most other networks. However, if multiple transactions are all writing to the same shared object, they must execute in sequential order. There is a limit to how many transactions that touch one specific object the network can process.
```
Don't use "we" or "us"

**Congestion control** is the system that limits the rate of transactions writing to a single shared object, preventing the network from becoming overloaded with checkpoints that take too long to execute.

> If you see transactions fail with the error `ExecutionCancelledDueToSharedObjectCongestion`, you are observing congestion control at work! Read this document to learn:

Reviewer 2: ```suggestion
If you see transactions fail with the error `ExecutionCancelledDueToSharedObjectCongestion`, you are observing object-based local fee markets at work. Continue reading to learn:
```

- How transaction space is allocated on a congested shared object.
- How to bid for priority access to limited space.
- How to structure your applications and transactions for maximum throughput.

## How congestion control works

Sui's congestion control algorithm runs every time a new batch of sequenced transactions is received from consensus.

> 1. All transactions are **sorted in order of gas price**.

Reviewer 2: ```suggestion
1. All transactions are **sorted in order of gas price**.
```
Is there an order to these? If not, should use ul using dashes instead of asterisks. And no need to decorate text anywhere in this doc.

1. Each transaction's **execution cost is estimated** using the [`TotalGasBudgetWithCap`](https://github.com/MystenLabs/sui/blob/main/crates/sui-core/src/authority/shared_object_congestion_tracker.rs#L143) heuristic.

1. The [`SharedObjectCongestionTracker`](https://github.com/MystenLabs/sui/blob/main/crates/sui-core/src/authority/shared_object_congestion_tracker.rs) keeps a running tally of how much **per-object congestion budget** is used. If *all* shared objects used by the transaction have enough budget left, the transaction is scheduled. Otherwise it is **deferred** until the next commit.

> 1. When a transaction is scheduled, it consumes budget for the *mutable* shared objects that it uses. (*Immutable* shared objects don't consume any budget, because multiple immutable uses of a shared object can be executed in parallel.)

Reviewer: This feels like an add-on to the previous point rather than a standalone bullet (maybe right after "how much per-object congestion budget issued")

### Priority ordering of transactions

> Transaction priority is determined solely by **gas price**. If you want your transaction to be assigned access to shared objects ahead of others, you must pay a higher gas price than the others. This is the only way to get priority access.

Reviewer: Maybe a good place to include a note about SIP45 and the 5xRGP amplification threshold. It's a bit of a type error because congestion control is not technically related to SIP45 but we don't talk about this elsewhere, and this docs seems like the place where folks are most likely to look.

Author: Can you propose any wording? I don't know exactly how that works

Reviewer: Here's my attempt (and tagging in @mwtian):

- When deciding which transactions to include in a consensus commit, Sui sorts by **gas price** (highest first)
- In addition, if a transaction's gas price is >=5X the current **reference gas price**, the transaction will be submitted to consensus faster than a tx submitted at the same time with a price <5X times the current reference gas price (see [SIP45](https://github.com/sui-foundation/sips/blob/main/sips/sip-45.md) for details)

Thus, if you want a transaction to be processed as quickly as possible, set a price that is at least 5X the RGP.

Reviewer 3: Since this discussion has been resolved, I assume we no longer want to add the details. But if we want to, we can just add a `NOTE: using a higher gas price (> 5 RGP) will also increase the likelihood the transaction gets included in the earliest consensus commit possible.`.

### Cost estimation of transactions

:::note

> We are actively developing an improved algorithm for estimating transaction execution cost, which we hope will significantly increase capacity for most use cases. This document describes how the current algorithm works.

Reviewer 2: I'd avoid making predictions.

:::

> Sui limits the per-commit execution capacity of each shared object. If transactions touching a shared object are smaller, more of them can fit. If they're bigger, not as many will fit.

Reviewer: ```suggestion
Sui limits the per-commit execution capacity of each shared object. If transactions touching a shared object have a low estimated cost, more of them can fit. If they have a high estimated cost, not as many will fit.
```

How is the estimated cost of a transaction determined for congestion control purposes? We use a heuristic called [`TotalGasBudgetWithCap`](https://github.com/MystenLabs/sui/blob/main/crates/sui-core/src/authority/shared_object_congestion_tracker.rs#L143).

> - A transaction's initial estimated cost is simply its **gas budget**. (Unfortunately, the final gas price of a transaction is not known until after execution, which is too late for us to use in scheduling.)

Reviewer: ```suggestion
- A transaction's initial estimated cost is simply its **gas budget**. (Unfortunately, the final gas cost of a transaction is not known until after execution, which is too late for us to use in scheduling.)
```

- For transactions with a very high gas budget, the estimated cost is then lowered to an upper limit that is determined based on the number of `MoveCall`s and the number of Move `Input`s in each call.

> This means that **if you want to make your transaction consume less space on a mutable shared object**, you must lower its gas budget, reduce the number of `MoveCall`s and Move `Input`s, or both.

Reviewer: I might move this to the last section, and title it "How to structure your applications and transactions for maximum throughput" (I liked that in your outline above)

### Transaction deferral

If a transaction is not able to be scheduled because it's trying to use a congested shared object with no space left, it will be **deferred**. This means that validators will hold onto the transaction try again to schedule it with the next commit, each time prioritizing all waiting transactions by gas price.

If a transaction has been deferred for several commits without successfully being scheduled, we cancel it with the `ExecutionCancelledDueToSharedObjectCongestion` error. You'll need to try it again with a higher gas price, or at a time when the shared objects it depends on are not as congested.

> ### Determining total per-object capacity

Reviewer: (assuming this section is repurposed to be filled with advice), I think the most actionable things are:
- Use a higher gas price for priority access [do we have advice on how to determine how high to go]
- Set a gas budget that is close to your actual execution cost [and mention how to dry run to determine this, caution that some txes have variable cost]
- If your object `O` is congested, optimize the gas usage of the functions that are commonly used to access the object. E.g., if all the accesses to your object happen via Move function `f`, halving the gas usage of `f` will effectively double the number of transactions that can touch `O` in a single checkpoint (modulo gas bucketing)
- When designing Move packages, avoid using a single shared object if possible --e.g., a DEX with a single object + dynamic fields for each currency pair is bad, an object per currency pair is good

Author: I just called it "practical takeaways" since not all of the suggestions improve throughput, lmk what you think

Reviewer: That looks great, thanks!

---

Congestion control is the system that limits the rate of transactions writing to a single shared object, preventing the network from becoming overloaded with checkpoints that take too long to execute.

The Sui network’s object-based architecture allows processing many different user transactions massively in parallel, in a way that’s not possible on most other networks. However, if multiple transactions are all writing to the same shared object, they must execute in sequential order. There is a limit to how many transactions the network can process that touch one specific object.

If you see transactions fail with the error `ExecutionCancelledDueToSharedObjectCongestion`, you are observing congestion control at work! Continue reading to learn:
- How transaction space is allocated on a congested shared object.
- How to bid for priority access to limited space.
- How to structure your applications and transactions for maximum throughput.

## How congestion control works

Sui's congestion control algorithm runs every time a new batch of sequenced transactions is received from consensus.

1. All transactions are **sorted in order of gas price**.

> 1. Each transaction's **execution cost is estimated** using the [`TotalGasBudgetWithCap`](https://github.com/MystenLabs/sui/blob/main/crates/sui-core/src/authority/shared_object_congestion_tracker.rs#L143) heuristic.

Author: I don't think so because if this file disappears the doc will be out of date and wrong, whether or not the link goes to an old version of a deleted file or a 404. It seemed better to at least link to the current version of the file in the meantime

1. The [`SharedObjectCongestionTracker`](https://github.com/MystenLabs/sui/blob/main/crates/sui-core/src/authority/shared_object_congestion_tracker.rs) keeps a running tally of how much **per-object congestion budget** is used. If *all* shared objects used by the transaction have enough budget left, the transaction is scheduled. In that case, it consumes budget for the *mutable* shared objects that it uses. (*Immutable* shared objects don't consume any budget, because multiple immutable uses of a shared object can be executed in parallel.) Otherwise it is **deferred** until the next commit.

### Priority ordering of transactions

> Transaction priority is determined solely by **gas price**. If you want your transaction to be assigned access to shared objects ahead of others, you must pay a higher gas price than the others. This is the only way to get priority access.

Author: done

### Determining total per-object capacity

At the protocol level, Sui is configured with a fixed per-commit limit and burst capacity for each shared object. On average, a shared object's activity cannot exceed the per-commit limit. However, the limit may be exceeded temporarily by short bursts of traffic.

There is no way to increase the total execution capacity of a particular shared object. That limit is set by the network to ensure that validators, full nodes, and indexers can all execute checkpoints in a reasonable amount of time.

## Practical takeaways

- For priority access to a congested shared object, set a higher gas price.

- Set a gas budget that is close to your actual execution cost. You can use dry runs to estimate this. Be aware that some transactions have variable cost.

- If your object `O` is congested, optimize the gas usage of the functions that are commonly used to access the object. For example, if all the accesses to your object happen via Move function `f`, halving the gas usage of `f` will effectively double the number of transactions that can touch `O` in a single commit (assuming the gas budget is correspondingly decreased).

> - When designing Move packages, avoid using a single shared object if possible. For example, a DEX application with a single, main shared object and dynamic fields for each currency pair would suffer much more congestion than one with a separate object per currency pair.

Author: I put the first two at the top because they don't require you to own a package in order to do them. I bumped this one up to third, so it's at the top of suggestions that do require you to own it
